pkgs/sops-install-secrets/darwin.go

```diff
@@ -24,22 +24,35 @@ func RuntimeDir() (string, error) {
 }
 
 func SecureSymlinkChown(symlinkToCheck string, expectedTarget string, owner, group int) error {
-	// not sure what O_PATH is needed for anyways
-	fd, err := unix.Open(symlinkToCheck, unix.O_CLOEXEC|unix.O_SYMLINK|unix.O_NOFOLLOW, 0)
+	// The flag combination of `O_NOFOLLOW|O_PATH` is used on Linux systems to
+	// provide a reference to a symlink directly, instead of following the link,
+	// however on Darwin systems `O_PATH` is not available. The `O_SYMLINK` flag
+	// is instead used to get a file descriptor to the symlink itself, and not
+	// to follow the link to the target file.
+	fd, err := unix.Open(
+		symlinkToCheck,
+		unix.O_CLOEXEC|unix.O_SYMLINK,
+		unix.O_RDONLY)
 	if err != nil {
 		return fmt.Errorf("failed to open %s: %w", symlinkToCheck, err)
 	}
 	defer unix.Close(fd)
 
+	// Verify that the symlink points to the expected target file.
 	buf := make([]byte, len(expectedTarget)+1) // oversize by one to detect trunc
-	n, err := unix.Readlinkat(fd, "", buf)
+	n, err := unix.Readlink(symlinkToCheck, buf)
```

Comment:
The whole point of this function is to operate on a fd rather than a path that can change. However changing this function to Readlink would introduce a TOCTOU attack, at which point we could also not check it.

Reply:
However I am currently also wondering if we actually need to make sure that the symlink points to the right target.

Reply:
Maybe instead of `Readlinkat` we can at least make sure with fstat that this is a still symlink? I think the code was originally inspired by systemd-tmpfiles: https://github.com/systemd/systemd/blob/fd2a114061786005f6f088652b100341165ef897/src/tmpfiles/tmpfiles.c#L2459

Reply:
Thanks for the review, it's much appreciated!

I suppose if we didn't check that the fd points to the correct target, then we might have a similar timing problem in between the creation of the symlink here https://github.com/Mic92/sops-nix/blob/master/pkgs/sops-install-secrets/main.go#L192 and the file open in `SecureSymlinkChown`.

Your fstat suggestion sounds reasonable to me, I'll make the changes and test it on my system.

Reply:
After looking over this a bit, `unix.Fstat` doesn't follow the symlink fd, so it doesn't seem like we'd be able to compare the files this way, unless I'm missing something.

I'll look into the subdirectory suggestion in the meanwhile.

Reply:
I've implemented the root-owned subdirectory approach. I've also replaced both of the Linux and Darwin implementations with this single implementation, however if you'd prefer to have the subdirectory approach be Darwin-only and leave the existing Linux implementation alone, I can revert some of the changes.

I think the latest build failed due to a buildbot infrastructure problem?